src/octoprint/util/comm.py

```diff
@@ -2321,6 +2322,16 @@ def busyIntervalSet():
                         elif action_name == "disconnect":
                             self._log("Disconnecting on request of the printer...")
                             self._callback.on_comm_force_disconnect()
+                        elif action_name == "shutdown":
+                            sarge.run(
+                                settings().get(
+                                    ["server", "commands", "systemShutdownCommand"]
+                                ),
```

Comment:
This can be unset, in which case this call would error out and cause the whole loop to error out. That's definitely not what we want, needs error handling + check for preconditions.

And  calling system commands from within the comm layer is also something that somehow makes me queasy. It might be time to introduce a manager type of class responsible for handling the restart OctoPrint/restart system/shutdown system commands and rather use that here. That then could also take care of error handling at a central place, and be reused in other places too (system API, software update and plugin manager plugin come to mind).

Reply:
I can definitely add the error handling.

I may not be the right person to make a manager class although I am willing to try. My python experience is limited.

Reply:
In any case, I share your queasiness about directly calling sarge and system commands form the inner loop. 

As for input, if you point me to a good class to take as example for the system command wrapper, I will give it a shot. If I have questions, I will come to the #dev-core discord channel.

Reply:
Just remember that you might also want to use [octoprint.util.commandline](https://github.com/OctoPrint/OctoPrint/blob/master/src/octoprint/util/commandline.py) for this. Maybe.